### PR TITLE
Implement forecast decomposition for ARIMA-like models

### LINKED ISSUE #1149: Implement forecast decomposition for ARIMA-like models
### 🚀 Feature Request

Implement forecast decomposition for `AutoARIMAModel` and `SARIMAXModel`

### Proposal

In `_SARIMAXBaseAdapter` implement methods:
```
def predict_components(df: pd.DataFrame) -> pd.DataFrame
def forcast_components(df: pd.DataFrame) -> pd.DataFrame
```
Details:
- implement in-sample decomposition in `predict_components` using [`ARIMAResults`](https://www.statsmodels.org/dev/generated/statsmodels.tsa.arima.model.ARIMAResults.html#statsmodels.tsa.arima.model.ARIMAResults)
- implement out-of-sample decomposition in `forcast_components` using `PredictionResults.results`
- estimate components using state-space representation of SARIMAX
- decompose prediction to SARIMA and exogenious contributions
- use `obs_intercept` to estimate exog components when `mle_regresion` is set
- name for SARIMA component in `SARIMAXModel`:  `sarima(p={},d={},q={},P={},D={},Q={})`
- name for ARIMA component in `AutoARIMAModel`:  `arima(p={},d={},q={})`
- exogenious components should be pefixed with `target_component_`

### Test cases

For `AutoARIMAModel` and `SARIMAXModel`

- Test that components sum up to target
- Test that the components names are correct for all possible parameters combination

### Additional context

See implementation in ETNA-1480 folder

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1149

### CODE REVIEW

File: tests/test_models/test_sarimax_model.py [hunk 1]
```diff
@@ -133,3 +135,103 @@ def test_forecast_1_point(example_tsds):
...
+    train, test = dfs_w_exog
+    pred_df = train if in_sample else test
+
+    model = _SARIMAXAdapter(trend=trend)
+    model.fit(train, regressors)
+
+    components_method = getattr(model, components_method_name)
+    components = components_method(df=pred_df)
+
+    assert set(components.columns) == set(expected_components)
```
Comment (alex-hse-repository) [inline tests/test_models/test_sarimax_model.py:182]: What if we have duplicated columns? Better to compare sorted versions

File: tests/test_models/test_autoarima_model.py [hunk 2]
```diff
@@ -144,3 +145,32 @@ def test_forecast_1_point(example_tsds):
 def test_save_load(example_tsds):
     model = AutoARIMAModel()
     assert_model_equals_loaded_original(model=model, ts=example_tsds, transforms=[], horizon=3)
+
+
+@pytest.mark.parametrize(
```
Comment (alex-hse-repository) [inline tests/test_models/test_autoarima_model.py:150]: We actually did't touch this adapter, I guess it is not really necessary. If you thing that it is useful also add test for components sum

File: etna/models/sarimax.py [hunk 3]
```diff
@@ -212,6 +214,126 @@ def get_model(self) -> SARIMAXResultsWrapper:
...
+
+        Returns
+        -------
+        :
+            dataframe with forecast components
+        """
+        fit_results = self._fit_results
+
+        model = fit_results.model
+        if model.hamilton_representation:
```
Comment (alex-hse-repository) [inline etna/models/sarimax.py:313]: Don't see the test for this))

File: etna/models/sarimax.py [hunk 4]
```diff
@@ -212,6 +214,126 @@ def get_model(self) -> SARIMAXResultsWrapper:
...
+        self._check_df(df, horizon)
+
+        exog_future = self._select_regressors(df)
+
+        forecast_results = fit_results.get_forecast(horizon, exog=exog_future).prediction_results.results
+        state = forecast_results.predicted_state[:, :-1]
+
+        if model.mle_regression:
+            components = self._mle_regression_decomposition(
+                state=state, ssm=forecast_results.model, exog=exog_future.values  # type: ignore
```
Comment (alex-hse-repository) [inline etna/models/sarimax.py:327]: Why do you ignore type here?

File: etna/models/sarimax.py [hunk 5]
```diff
@@ -212,6 +214,126 @@ def get_model(self) -> SARIMAXResultsWrapper:
         """
         return self._fit_results
 
+    @staticmethod
+    def _prepare_components_df(components: np.ndarray, model: SARIMAX) -> pd.DataFrame:
```
Comment (alex-hse-repository) [inline etna/models/sarimax.py:218]: Don't see tests for this method

Comment (alex-hse-repository) [inline etna/models/sarimax.py:218]: Is `SARIMAX` a correct typing as we also expect to use it for `AutoSARIMAX`?

File: etna/models/sarimax.py [hunk 6]
```diff
@@ -212,6 +214,126 @@ def get_model(self) -> SARIMAXResultsWrapper:
...
+        return design_mat
+
+    def _mle_regression_decomposition(self, state: np.ndarray, ssm: SimulationSmoother, exog: np.ndarray) -> np.ndarray:
+        """Estimate SARIMAX components for MLE regression case."""
+        design_mat = self._prepare_design_matrix(ssm)
+
+        components = np.sum(design_mat * state, axis=1).T
+
+        if len(exog) > 0:
+            exog_params = np.linalg.lstsq(a=exog, b=np.squeeze(ssm["obs_intercept"]))[0]
```
Comment (alex-hse-repository) [inline etna/models/sarimax.py:249]: May you reference sth here to see the details of this method. The same for `_state_regression_decomposition`

File: etna/models/sarimax.py [hunk 7]
```diff
@@ -6,10 +6,12 @@
 from typing import Sequence
```
Comment (alex-hse-repository) [inline etna/models/sarimax.py:6]: In `SARIMAX` and `AutoSARIMAX` add notes about the target components

Comment (brsnw250) [reply etna/models/sarimax.py:6]: Added

File: etna/models/sarimax.py (hunk 5 above)
Comment (brsnw250) [reply etna/models/sarimax.py:218]: Here we only add names to components and convert to dataframe, so this functionality is tested in `test_components_names`

Comment (brsnw250) [reply etna/models/sarimax.py:218]: Yes,  `pmdarima.auto_arima` fits SARIMAX model

File: tests/test_models/test_autoarima_model.py (hunk 2 above)
Comment (brsnw250) [reply tests/test_models/test_autoarima_model.py:150]: Yes, we can remove this test

File: tests/test_models/test_sarimax_model.py [hunk 8]
```diff
@@ -133,3 +135,103 @@ def test_forecast_1_point(example_tsds):
...
+
+
+@pytest.mark.long_2
+@pytest.mark.parametrize(
+    "components_method_name,in_sample", (("predict_components", True), ("forecast_components", False))
+)
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_models/test_sarimax_model.py:190]: Yes, they affect how model is estimated inside. Some do additional preprocessing. We should ensure that decomposition is correct for all this cases.

File: etna/models/sarimax.py (hunk 3 above)
Comment (brsnw250) [reply etna/models/sarimax.py:313]: `test_decomposition_hamiltonian_repr_error` in `test_sarimax_model.py`

File: etna/models/sarimax.py (hunk 6 above)
Comment (brsnw250) [reply etna/models/sarimax.py:257]: Added comments

File: etna/models/sarimax.py (hunk 4 above)
Comment (alex-hse-repository) [reply etna/models/sarimax.py:347]: Sounds quite implicit, may you  add comments about this

Comment (brsnw250) [reply etna/models/sarimax.py:349]: Added

**Review by alex-hse-repository — APPROVED** [review]

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**